### PR TITLE
WIP add helper for using pb-rs from build.rs scripts

### PR DESCRIPTION
**Pull request by nerdrew**
Thoughts on adding a helper to make build.rs scripting easier?

### CODE REVIEW

File: pb-rs/src/lib.rs [hunk 1]
```diff
@@ -10,3 +10,109 @@ pub mod keywords;
...
 pub mod scc;
 pub mod types;
+
+use errors::Error;
+use types::{Config, FileDescriptor};
+use std::path::PathBuf;
+use failure::ResultExt;
+
+#[derive(Debug, Default)]
+pub struct CompileBuilder {
```
Comment (tafia) [inline pb-rs/src/lib.rs]: I would prefer having a `ConfigBuilder` only. There is nothing *compiled* in my opinion.

File: pb-rs/src/lib.rs [hunk 2]
```diff
@@ -10,3 +10,109 @@ pub mod keywords;
...
+    out_file: Option<PathBuf>,
+    include_paths: Vec<PathBuf>,
+    single_module: bool,
+    no_output: bool,
+    error_cycle: bool,
+    headers: bool,
+}
+
+impl CompileBuilder {
+    pub fn new(in_files: Vec<PathBuf>, output: Option<PathBuf>, output_dir: Option<PathBuf>, mut include_paths: Vec<PathBuf>) -> Result<CompileBuilder, ::failure::Error> {
```
Comment (tafia) [inline pb-rs/src/lib.rs]: Can you add comments for all these fns?

File: pb-rs/src/lib.rs (hunk 1 above)
Comment (nerdrew) [reply pb-rs/src/lib.rs]: Sure. I'll rename it to `ConfigBuilder`.

**Comment by nerdrew** [comment]
> I love the idea of using a builder. I would prefer limiting it to a config builder though ... and still have a FileDescriptor::run called.

I'm not sure what you mean by `FileDescriptor::run`. Are you thinking a new method that accepts a slice of `Config`s? Or perhaps the `ConfigBuilder` has an iter you can call to iterate over the generated configs and then call `FileDescriptor::write_proto`?

**Comment by nerdrew** [comment]
> Could you also apply rustfmt (rustup component add rutfmt-preview) and add comments to all these builder fn (so they appear in the pb-rs docs).

LOL. I think my version of `rustfmt` might be different than yours. I had to disable it for this project since it changed every file when I saved :). Ya, I'll run it on the code I add.

**Comment by nerdrew** [comment]
Something like?

```rust
impl ConfigBuilder {
    pub fn build(self) -> Vec<Config> {
        self.in_files
            .iter()
            .map(|in_file| {
                let mut out_file = in_file.with_extension("rs");

                if let Some(ref ofile) = self.out_file {
                    if ofile.is_dir() {
                        out_file = ofile.join(out_file.file_name().unwrap());
                    } else {
                        out_file = ofile.into();
                    }
                }

                Config {
                    in_file: in_file.to_owned(),
                    out_file: out_file,
                    import_search_path: self.include_paths.clone(),
                    single_module: self.single_module,
                    no_output: self.no_output,
                    error_cycle: self.error_cycle,
                    headers: self.headers,
                }
            }).collect()
    }
}
```

```rust
impl FileDescriptor {
    pub fn run(configs: &[Config]) -> Result<()> {
        for config in configs {
            Self::write_proto(&config)?
        }
        Ok(())
    }
}
```

```rust
FileDescriptor::run(&compiler.build()).map_err(|e| e.into())
```

**Comment by tafia** [comment]
yes if you don't mind

**Comment by nerdrew** [comment]
@tafia I updated this PR.

**Comment by tafia** [comment]
Thanks!
Could you please rustfmt again?

**Comment by nerdrew** [comment]
Ran: ```git diff --name-only upstream/master | grep -E '^pb-rs.*\.rs$' | xargs rustfmt```

File: pb-rs/src/types.rs [hunk 3]
```diff
@@ -1879,7 +1890,7 @@ fn split_package(package: &str) -> (&str, &str) {
     }
 }
 
-const MAGIC_HEADER: &'static str = "//! Automatically generated mod.rs";
+const MAGIC_HEADER: &'static str = "// Automatically generated mod.rs";
```
Comment (nerdrew) [inline pb-rs/src/types.rs]: This is unrelated to this PR, but was failing for me when using `include!`

Comment (tafia) [reply pb-rs/src/types.rs]: Can you rebase? I think this was already done in a previous PR.

**Comment by nerdrew** [comment]
My rustfmt:

```
% cargo +stable fmt --version
rustfmt 0.99.1-stable (da17b689 2018-08-04)
```

**Comment by nerdrew** [comment]
I ran `cargo +stable fmt` from the root. I just re-ran it and nothing changed.

**Comment by tafia** [comment]
Alright then, I am not sure what happened exactly.
Thanks!

**Pull request merged**